### DISCUSSION TITLE
Raspberry Pi VSG installation

### DISCUSSION THREAD

**Discussion by DaveAllerton** [vsg-dev/VulkanSceneGraph#1019] [Q&A] [answered]
I run flight simulation software using OSG and would like to port the software to VSG on a Raspberry Pi 4B. I ran update and upgrade on my RPi, so it is fully up to date. I also run lots of OpenGL v4 on the RPi. 

I meticuously followed the instructions for Unix from https://vsg-dev.github.io/vsgTutorial/SettingTheScene/BuildingVulkanSceneGraph
VulkanSceneGraph builds OK. Note, there is a typo export LD_LIBRARY_PATH=${D_LIBRARY_PATH}:${INSTALL_DIR}/lib (D_LIBRARY)
vsgExamples builds OK but when I run vsgdraw, I get
=========
~/Projects/vsgXchange>vsgdraw
terminate called after throwing an instance of 'vsg::Exception'
Aborted
=========
I also need vsgXchange because I use OpenFlight files (.flt) which I will need to convert. But when I run the cmake, I get
=========
~/Projects/vsgXchange>cmake . -DCMAKE_INSTALL_PREFIX=${INSTALL_DIR}
-- Reading 'vsg_...' macros from /home/pi/Install/lib/cmake/vsg/vsgMacros.cmake - look there for documentation
-- Could NOT find CURL (missing: CURL_LIBRARY CURL_INCLUDE_DIR) 
-- The following OPTIONAL packages have been found:

 * vsg (required version >= 1.0.10)
 * Freetype

-- The following REQUIRED packages have been found:

 * Vulkan (required version >= 1.1.70.0)
 * PkgConfig
 * Threads

-- The following OPTIONAL packages have not been found:

 * CURL

-- Configuring done
CMake Error in src/CMakeLists.txt:
  Imported target "assimp::assimp" includes non-existent path

    "/usr/lib/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
=========
sudo apt-get install curl says that curl is installed and up to date

Any suggestions? I can't find any links to anyone who has build and run Vulkan examples on a Raspberry Pi, although there seems to be claims that it has been done. I've still got to convert 1500 lines of OSG to VSG so any help to get past the installation phase would be appreciated.

**Comment by timoore** [comment]
I don't have any experience with the Rasberry Pi, but...

Can you run any Vulkan programs, such as vulkaninfo or vkcube that comes with the SDK? If so, run vsgdraw under gdb and see what the problem is.

How did you install assimp? What version?

**Reply by DaveAllerton** [reply]
This is what I see when I run vulkinfo and vkcube.

=========
~/Projects/vsgXchange>vulkaninfo
Cannot create Vulkan instance.
This problem is often caused by a faulty installation of the Vulkan driver or attempting to use a GPU that does not support Vulkan.
ERROR at /build/vulkan-tools-TruZ6I/vulkan-tools-1.2.162.0+dfsg1/vulkaninfo/vulkaninfo.h:666:vkCreateInstance failed with ERROR_INCOMPATIBLE_DRIVER

~/Projects/vsgXchange>vkcube
vkEnumerateInstanceExtensionProperties failed to find the VK_KHR_surface extension.

Do you have a compatible Vulkan installable client driver (ICD) installed?
Please look at the Getting Started guide for additional information.
=========
I did not install the SDK because the tutorial implies it is not needed with the main install. Perhaps I misunderstood.

**Reply by DaveAllerton** [reply]
libassimp-dev is version 5.0.1-ds0-2
I have also installed assimp-utils

**Comment by robertosfield** [comment, marked as answer]
A number of questions to answer, which makes for a difficult to follow for others.  Could you post one topic per thread in future to help other more easily find/read/reply. Thanks.

I'll reply with one topic per post now to avoid the shush of issues causing more tangled thread.

---

With the typo, does the follow diff look fix the error(s) you've spotted?  Simply changing ${D_LIBRARY_PATH} to ${LD_LIBRARY_PATH}

~~~ sh
$ git diff
diff --git a/1_SettingTheScene/BuildingVulkanSceneGraph.md b/1_SettingTheScene/BuildingVulkanSceneGraph.md
index 669533c..fb4e836 100644
--- a/1_SettingTheScene/BuildingVulkanSceneGraph.md
+++ b/1_SettingTheScene/BuildingVulkanSceneGraph.md
@@ -19,7 +19,7 @@ First we'll set up some environment variable, for ease of use you may wish to pu
 export INSTALL_DIR=~/Install
 export PROJECT_DIR=~/Projects
 export PATH=${PATH}:${INSTALL_DIR}/bin
-export LD_LIBRARY_PATH=${D_LIBRARY_PATH}:${INSTALL_DIR}/lib
+export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${INSTALL_DIR}/lib
 
 # env vars that are used to find example data, and locating where to cache files downloaded from HTTP during database paging
 export VSG_FILE_PATH=${PROJECT_DIR}/vsgExamples/data
~~~

I have checked this into vsgTutorial, if there are further typos let me know, or just submit a PR to the appropriate repo.

**Reply by DaveAllerton** [reply]
Correct, that was the typo. My errors occured after I corrected the typo.

**Comment by robertosfield** [comment]
For OpenFlight support you'll need to use osg2vsg which you'll need to build the OSG, then osg2vsg, then build vsgXchange as it will then pull in osg2vsg if it's available.  This is quite a few hoops to jump through unfortunately as Assimp doesn't have a OpenFlight loader.

Longer term I'd like to see an OpenFlight loader implemented for the VSG itself, rather than required osg2vsg.  osg2vsg also needs a rewrite as was written in the early days of the VSG project and doesn't take advantage of all the modern VSG features like vsg::ShaderSet/GraphicsPipelineConfigurator to create subgraphs.

The vsgXchange::assimp loader does use vsg::ShaderSet/GraphicsPipelineConfigurator so is more flexible, extensible and supports all the modern features like PBR, lighting and shadows.

Another route for the OpenFlight support would be to build osg2vsg and then convert the .flt files to .vsgb and then load the .vsgb files in your flight simulator application.  This would avoid need to link to the OSG/osg2vsg at runtime.

As for getting things working on the Raspberry Pi, I don't have one myself, but I know others that are using the VSG on it.  Hopefully they'll chip in here with how to get things to run smoothly.

The vsg::Exception being thrown by vsgdraw is probably down to the Vulkan instance or device failing to be created.  This would most likely a driver setup issue.  Try running:

~~~ sh
vsgviewer vsgExamples/data/models/lz.vsgt 
~~~

As it'll catch the vsg::Exception and report the message to the console. Running _vsgdeviceselection_ should also report more info.

**Reply by DaveAllerton** [reply]
Thanks, you suggestions to move from OSG to VSG are along the lines I intend to follow.

This is the output from vsgviewer
======
vsgviewer vsgExamples/data/models/lz.vsgt 
[Exception] - Error: vsg::Instance::create(...) failed to create VkInstance. result = -7
======
I get a similar error running vkdeviceselection. The RPi runs the Raspberry Pi OS (not Raspian) and uses Wayland rather than X11.

**Reply by robertosfield** [reply]
You'll need to get standard Vulkan applications like vulkaninfo running first, then once you've sorted out the driver issues move on to the VSG examples.  

FYI,  VkResult value -7 is VK_ERROR_EXTENSION_NOT_PRESENT:

The Vulkan docs on VkResult:
https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkResult.html

**Reply by reedev** [reply]
and that means that the drivers are not installed, see my earlier suggestion

**Comment by reedev** [comment]
Hi,

I am running VSG based apps, including assimp loaded models, on an rpi 4 so it is possible!

I would like to help to get the VSG up and running on your side. Let's take it step by step and could you provide means to reproduce your issue? What OS, version, etc are you using? I am using Raspbian 11 64 bit

About curl, you need to install the dev package, I install libcurl4-openssl-dev

**Reply by DaveAllerton** [reply]
I am running the current Raspberry Pi OS (not Raspian) which is based on Debian. It is a 64-bit implementation.
sudo apt-get install curl says it is already the current version. 
I installed livbcurl4-openssl-dev and now cmake finds CURL but I still get the error message
========
-- Configuring done
CMake Error in src/CMakeLists.txt:
  Imported target "assimp::assimp" includes non-existent path

    "/usr/lib/include"
========

**Reply by reedev** [reply]
Since I experienced issues with assimp from the repo, I am building and installing assimp from source code

**Comment by reedev** [comment]
you also need to install package mesa-vulkan-drivers on raspbian 11

**Reply by DaveAllerton** [reply]
I installed the mesa-vullkan drivers and it now runs vsgdraw (draws some terrain) - good news, but I get the error message
=========
~/Projects/vsgExamples>vsgdraw
WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.
~/Projects/vsgExamples>
=========

**Reply by robertosfield** [reply]
The VSG doesn't emit this warning, so it'll be the driver itself flagging itself.

I don't know what drives are available on the Pi so I'll have to defer to others and yourself.  It may be that the driver is fine and over time will become conformant and this issue could go away, or perhaps other drivers will be available work seamlessly.

W.r.t Xcb vs Wayland, I implemented src/vsg/platform/xcb/Xcb_Window.cpp early in the VSG project as that's what most prevalent on Linux/Vulkan examples back in 2018.  I have looked at Wayland/Vulkan implementations in the past year and it looks relatively straight forward to implement a Wayland_Window.cpp,  I haven't had the free time to implement it, but if no one else gets to it I will eventually get round to it.  Could be a while though as my current tasks list is rather full.

**Reply by reedev** [reply]
fwiw I always see this warning too but so far I have not experienced any issue. But I have to say that I am not yet using advanced shaders or so

**Reply by DaveAllerton** [reply]
Thanks for the tips.
I can now build and run vsgviewer with models/lz.vsgt. I tried to build vsgXconv but cmake complained about the folder /usr/libs/include not existing, so I created it and cmake ran to completion. But "make -j 8" gives the following error, 86% through the build:

=====
[ 86%] Building CXX object src/CMakeFiles/vsgXchange.dir/curl/curl.cpp.o
/home/pi/Projects/vsgXchange/src/assimp/assimp.cpp: In member function ‘vsg::ref_ptr<vsg::Node> SceneConverter::visit(const aiScene*, vsg::ref_ptr<const vsg::Options>, const vsg::Path&)’:
/home/pi/Projects/vsgXchange/src/assimp/assimp.cpp:721:31: error: ‘const struct aiScene’ has no member named ‘mName’
  721 |     std::string name = scene->mName.C_Str();
      |                               ^~~~~
[ 90%] Building CXX object src/CMakeFiles/vsgXchange.dir/openexr/openexr.cpp.o
make[2]: *** [src/CMakeFiles/vsgXchange.dir/build.make:173: src/CMakeFiles/vsgXchange.dir/assimp/assimp.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:218: src/CMakeFiles/vsgXchange.dir/all] Error 2
make: *** [Makefile:149: all] Error 2
=====
It would be helpful to run vsgXconv and osg2vsg to convert my OSG files to VSG files. Then I can tackle rewriting my OSG renderer in VSG. I suspect the Raspberry Pi version of assimp is not the one I should be using.

**Reply by robertosfield** [reply]
This looks like an assimp version issue as it looks like mName isn't supported by the version of assimp you have on your system.  It may be possible to add a #ifdef assimp version check to disable the use.  Alternative you try updating or a more modern version of assimp.

**Reply by DaveAllerton** [reply]
I downloaded assimp from github and it builds correctly. I can now build and run vsgXconv. I will now install OSG and osg2svg and see how I go with translating OpenFlight .flt files and textures. Thanks for the suggestions.
